<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta content="text/html; charset=utf-8" http-equiv="Content-Type">
<title>Usage - InterARTIC Documentation</title>
<meta http-equiv="X-UA-Compatible" content="IE=edge">

<meta name="generator" content="mkdocs-1.1.2, mkdocs-gitbook-1.0.7">

<link rel="shortcut icon" href="../images/favicon.ico" type="image/x-icon">
<meta name="HandheldFriendly" content="true"/>
<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="black">
<meta rel="next" href="" />
<link href="../css/style.min.css" rel="stylesheet"> 
</head>

<body>
<div class="book">
<div class="book-summary">

<nav role="navigation">
<ul class="summary">
<li>
<a href=".." target="_blank" class="custom-link">InterARTIC Documentation</a>
</li>
<li class="divider"></li>
<li class="chapter" data-path="">
<a href="..">Home</a>
<li class="chapter" data-path="installation/">
<a href="../installation/">Installation</a>
<li class="chapter active" data-path="usage/">
<a href="./">Usage</a>
<li class="chapter" data-path="troubleshooting/">
<a href="../troubleshooting/">Troubleshooting</a>
<li class="divider"></li>



<li><a href="http://www.mkdocs.org">
Published with MkDocs
</a></li>

<li><a href="https://github.com/GitbookIO/theme-default">
Theme by GitBook
</a></li>
</ul>

</nav>

</div> <!-- end of book-summary -->

<div class="book-body">
<div class="body-inner">
<div class="book-header" role="navigation">

<!-- Title -->
<h1>
<i class="fa fa-circle-o-notch fa-spin"></i>
<a href="." ></a>
</h1>

</div> <!-- end of book-header -->

<div class="page-wrapper" tabindex="-1" role="main">
<div class="page-inner">

<section class="normal markdown-section">



<h1 id="usage">Usage</h1>
<div class="toc">
<ul>
<li><a href="#usage">Usage</a><ul>
<li><a href="#adding-a-job">Adding a job</a></li>
<li><a href="#parameters">Parameters</a><ul>
<li><a href="#input-folders">Input Folders</a><ul>
<li><a href="#input-directory-file-structure">Input directory file structure</a></li>
</ul>
</li>
<li><a href="#parameter-descriptions">Parameter Descriptions</a></li>
</ul>
</li>
<li><a href="#progress-page">Progress Page</a><ul>
<li><a href="#what-happens-if-an-error-occurs-during-the-run">What happens if an error occurs during the run?</a></li>
<li><a href="#what-happens-when-a-job-is-completed">What happens when a job is completed?</a></li>
</ul>
</li>
<li><a href="#output-page">Output Page</a><ul>
<li><a href="#files-produced">Files Produced</a></li>
<li><a href="#data-visualisation">Data Visualisation</a><ul>
<li><a href="#variants-found">Variants Found</a></li>
<li><a href="#plots-produced-from-pipeline">Plots Produced From Pipeline</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<h2 id="adding-a-job">Adding a job</h2>
<p>To begin the process of adding a job, click the 'Add Job' button located underneath the Jobs Queue on the home page. </p>
<h2 id="parameters">Parameters</h2>
<p>Input the necessary parameters (see Parameter Descriptions below). Parameters required for any type of job run are denoted with an asterix (*).</p>
<p>For the file path inputs, please enter absolute paths. See below for help.</p>
<h3 id="input-folders">Input Folders</h3>
<p>By default, the data input folder should be set up as:</p>
<pre><code>~user/
    data/
        input_folder1/                               
            name/
                uuid/
                    # input files here
        input_folder2/
            name/
                uuid/
                    # input files here

</code></pre>

<p>When prompted to select an input folder, click the drop down menu or type in the name of the input folder as in the data folder in the file tree above. For example, if you want to run the pipeline on <code>input_folder1</code>, type <code>input_folder1</code> into the search bar or select it in the drop down menu.</p>
<h4 id="input-directory-file-structure">Input directory file structure</h4>
<ul>
<li>You must rename each folder/ file in the input folder to: fast5_pass, fast5_fail, fastq_pass, fastq_fail, sequencing_summary.txt, sample-barcode.csv</li>
<li>If a single sample is being run through the pipeline:<ul>
<li>If available, the file path for a read file should be inputted. </li>
<li>If unavailable, the artic gather/demultiplex command will generate one.</li>
</ul>
</li>
<li>If multiple samples are being run through the pipeline:<ul>
<li>A CSV file containing the data's sample names and barcodes should be placed in the input folder and named ‘sample-barcode.csv’. </li>
<li>A sample CSV file is below:</li>
</ul>
</li>
</ul>
<pre><code>sample3,NB03
sample4,NB04
sample5,NB05
sample6,NB06
</code></pre>

<p>A sample file structure is as below:</p>
<pre><code>input_folder/
    fast5_pass/
        A_10.fast5
        A_11.fast5
    fast5_fail/
        A_0.fast5
        A_1.fast5
    fastq_pass/
        B_0.fastq
        B_1.fastq
    fastq_fail/
        B_10.fastq
        B_11.fastq
    sample-barcode.csv
    sequencing_summary.txt
    readfile.fastq                        #optional file
primer-schemes/
    nCoV-2019/
        V1/
            nCoV-2019.log
            nCoV-2019.pdf
            nCoV-2019.pickle
            nCoV-2019.reference.fasta
            nCoV-2019.reference.fasta.fai
            nCoV-2019.scheme.bed
            nCoV-2019.svg
            nCoV-2019.tsv
</code></pre>

<h3 id="parameter-descriptions">Parameter Descriptions</h3>
<p>You can customize the parameters by typing into the respective text box. </p>
<ul>
<li><strong>Job name:</strong> A unique name for your job, so you may identify your output files with it.</li>
<li><strong>Pipeline:</strong> Select the pipeline within ARTIC that you wish to run your data files through.</li>
<li><strong>Input folder:</strong> Your main data folder. <ul>
<li>This folder contains folders such as fast5_pass, fastq_pass, etc.</li>
</ul>
</li>
<li><strong>Read file:</strong> If you are inputting data files for a single sample run where you have a file ending in “.fastq” already made, input the file path for this read file. <ul>
<li>You may get this error in your output, but this can be ignored: Include screenshot of harmless error that might occur</li>
</ul>
</li>
<li><strong>Output folder:</strong> If your chosen output folder is already created, enter the file path for this folder. <ul>
<li>Otherwise, you may enter a file path to a folder that doesn’t exist yet. If you do this, ensure that the parent directory exists. <ul>
<li>For example, if you are inputting this file path as your output folder: path/to/file/hello/world, the folder “hello” must already exist for the “world” folder to be created.</li>
</ul>
</li>
<li>If this is not inputted, an output folder will be generated within the input folder.</li>
</ul>
</li>
<li><strong>Primer scheme folder:</strong> Enter the file path to your primer schemes. <ul>
<li>This is the folder containing, for example, the folder nCoV-2019 which contains the V1, V2, etc folders.</li>
</ul>
</li>
<li><strong>Primer scheme name:</strong> Enter the primer scheme name used for your nanopore sequencing run.<ul>
<li>Following a similar example to the previous parameter description, here you will enter a path such as nCoV-2019/V1.</li>
</ul>
</li>
<li><strong>Primer/Barcode type:</strong> Enter the type of primer/barcode used. <ul>
<li>Either select from the options available or enter the name of the primer/barcode in the text box. </li>
<li>This is only used for folder-naming purposes.</li>
</ul>
</li>
<li><strong>Minimum/Maximum length:</strong> If you selected from the available options in the primer/barcode type section, you may find the minimum and maximum length already filled out. <ul>
<li>If not, set this to the minimum/maximum length of your primers.</li>
</ul>
</li>
<li><strong>Thread usage/Normalise:</strong> Change the prefilled values if you wish.<ul>
<li>Please note that changing the threads/normalise values, they are changed globally for all commands.</li>
</ul>
</li>
<li><strong>Single or Multiple samples:</strong> Select the appropriate option.</li>
</ul>
<p>When you are confident that your parameter selections are correct, click on the “Submit Job(s)” button. You will be redirected to the progress page after clicking this button.</p>
<h2 id="progress-page">Progress Page</h2>
<p>The progress page displays the stream of standard output being produced by your job run. Here you can see which commands are currently running and any errors that occur. Each job run has its own progress page which can be accessed via the home page or parameters page by clicking on the job name in the jobs queue.</p>
<p>For each job, the progress page will display:</p>
<ul>
<li>The job name</li>
<li>The job's place in queue</li>
<li>An 'Abort Job' button</li>
<li>The overall progress of the job in the form of a progress bar and the number of steps remaining in the pipeline</li>
<li>A 'View Job Parameters here' button</li>
<li>The current standard output obtained from the job</li>
</ul>
<p>The 'View Job Parameters here' button, when clicked, will display the job's parameters that have been entered by the user.</p>
<p>The 'Abort Job' button can be used to terminate the job. A confirmation window will appear when you click on the abort button. If you continue, you will then be asked to confirm whether you wish to delete the files created by the job. After this, you will then be directed back to the home page.</p>
<h3 id="what-happens-if-an-error-occurs-during-the-run">What happens if an error occurs during the run?</h3>
<p>If an <strong>error</strong> occurs during a run, a <strong>red</strong> notification will appear. You can either let the job continue to run, or click the ‘Re-run’ button. Harmless errors sometimes occur in the ARTIC pipeline, so it may be worth waiting for the run to finish and then assessing your output.</p>
<p>Clicking the ‘Re-run’ button will allow you to abort the currently running job and re-run the job with editted parameters.</p>
<p>A confirmation window will appear when you click on the ‘Re-run’ button asking you to confirm that you wish to abort the current job and whether to delete the files created by the job. </p>
<p>You will then be redirected to the parameters page where the information from the job in question will be automatically filled in. </p>
<p>You can make any changes necessary, and the new job will be added to the end of the queue following submission. </p>
<h3 id="what-happens-when-a-job-is-completed">What happens when a job is completed?</h3>
<p>When a job is completed, a ‘Go to Output’ button will appear at the top of the page. Click the button to be redirected to the output page. The job will also be moved to the 'Completed Jobs' list on the home page where you can click on the job name and be redirected to the output page for that job.</p>
<h2 id="output-page">Output Page</h2>
<p>The Output Page has two main sections, one for the files produced during the run, and the other for data visualisation to enable a fast quality check of the sample. At the bottom of the page, there is a 'Go to Progress' button which will redirect you to the progress page of the job if you click on it.</p>
<h3 id="files-produced">Files Produced</h3>
<p>This section contains a list of the files produced from the job, along with the location in which they were saved. This location should be the output folder that you inputted on the parameters page, or if you did not input one, it will be a folder named “output” created inside your input folder.</p>
<h3 id="data-visualisation">Data Visualisation</h3>
<p>The “Data Visualisation” section comprises two main parts.</p>
<ul>
<li>Variants found</li>
<li>Plots produced from pipeline</li>
</ul>
<h4 id="variants-found">Variants Found</h4>
<p>This section produces a simple graph/s from the data inside the <code>&lt;sample_name&gt;.pass.vcf.gz</code> file/s produced by the pipeline. If multiple samples are selected, multiple files will be produced (assuming all runs are successful).</p>
<p>The 'Produce Graphs' button creates the graph/s. The vertical lines on the graph represent a variant, with the specific numerical position of the variant on the bottom axis. The mutation that occurred is labelled REF -&gt; ALT, meaning that the nucleotide/s in the reference genome are on the left of the arrow, and the variant is on the right of the arrow. The height of the vertical line indicates the read depth of the specific variant. To download the graph, click on the 'Download' hyperlink in the lower-left corner of the graph of interest.</p>
<p>These graphs, as mentioned, require interARTIC to read the <code>&lt;sample_name&gt;.pass.vcf.gz</code> file/s from the output folder. If there are no such files found, then no graphs will be produced. If you would not like the web app to access this, please click the “Disable” option under “Enable VCF graphs to be generated?” and click the 'Confirm' button. When you are no longer given the option to produce the graphs, and the status of the “Enable VCF graphs to be generated?” section is “Disabled” this is complete. If you would like to re-enable the graphs to be produced, just select “Enable” and once again confirm. The functionality should be restored.</p>
<p>If no <code>&lt;sample_name&gt;.pass.vcf.gz</code> files are found in the output folder, the message “Vcf graph could not be made: No pass.vfc.gz file/s found in the output folder.” will be displayed. As no files of the suitable format have been found, these graph/s cannot be produced. This may be due to errors or problems during the pipeline, so checking error messages in the progress page's standard output section is important. </p>
<h4 id="plots-produced-from-pipeline">Plots Produced From Pipeline</h4>
<p>This section enables you to preview the mean amplicon read depth plots produced from the pipeline. The bar plot shows the mean amplicon read depth over the amplicons, while the box plot shows the mean amplicon read depth over the read groups.</p>
<p>These plots are accessed from the output folder, and require a copy of them to be saved into the interARTIC web app’s folder to enable previewing. If you would not like this to happen, please disable the feature by selecting the “Disable” option under the “Would you like the plots to be previewed?” section, and clicking the 'Confirm' button. The status should be updated to say “Disabled” and the plots will be automatically removed from the interARTIC web app's folder. To re-enable previewing, select “Enable” and press the 'Confirm' button again, and the option to preview graphs should be restored. </p>
<p>If no barplot.png or boxplot.png files are found in the output folder, the message “Plots cannot be previewed: No plots were found in the output folder.” will be displayed. As no files in suitable format have been found, these graph/s cannot be previewed. This may be due to errors or problems during the pipeline, so checking error messages in the progress page's standard output section is important.</p>


</section>

</div> <!-- end of page-inner -->
</div> <!-- end of page-wrapper -->

</div> <!-- end of body-inner -->

</div> <!-- end of book-body -->
<script src="../js/main.js"></script>
<script src="../js/gitbook.min.js"></script>
<script src="../js/theme.min.js"></script>
</body>
</html>